src/asyncProfiler.cpp

```diff
@@ -150,6 +150,16 @@ void Profiler::storeMethod(jmethodID method) {
 }
 
 void Profiler::recordSample(void* ucontext) {
+    // Check deadline
+    if (time(NULL) > _deadline) {
+        char error[] = "Disabling profiler due to deadline\n";
+        write(STDERR_FILENO, error, sizeof(error));
+
+        _running = false;
+
+        setTimer(0, 0);
```

Comment:
It seems to be true, because `alarm()` [can be safely called inside a signal handler](http://man7.org/linux/man-pages/man7/signal.7.html) and [it is implemented](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/posix/alarm.c;h=f7d97f1f330d3c698e27514616826a149b666315;hb=bd2260a2069993cdb14cffcbb2b0dc04b9888aee) using `setitimer()`.